text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,648 @@
+the application code. Instead, callbacks or handlers are registered and then
+called by messaging SDKs to forward messages to the application.
+
+A "Deliver" span covers the call of such a callback or handler and should link
+to the creation context of all messages that are forwarded via the respective
+call.
+
+#### Instrumenting pull-based scenarios
+
+In pull-based consumer scenarios, the delivery of messages is requested by the
+application code. This usually involves a blocking call, which returns zero or
+more messages on completion.
+
+A "Receive" span covers such calls and should link to the creation context of
+all messages that are forwarded via the respective call.
+
+#### General considerations for both push-based and pull-based scenarios
+
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
+might pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the creation context of the messages forwarded, thus those
+spans can link to zero, one, or multiple producer spans.
+
+For single-message scenarios, and if the "Deliver" or "Receive" spans would be
+root spans of a new trace, the creation context may also be used as a parent on
+those operations in addition to being added as a link. Keeping single-messages
+operations in the same trace can greatly improve the user experience.
+
+#### Settlement
+
+Messages can be settled in a variety of different ways:
+
+* Settlement happens on the broker.
+* Settlement operations are triggered manually by the user.
+* In callback scenarios settlement can be automatically triggered by messaging SDKs
+  based on return values of callbacks.
+
+A "Settle" span should be created for every settlement operation, no matter
+which party triggered it.  SDKs will, in some cases, auto-settle messages in
+push-scenarios when messages are delivered via callbacks. In cases where it is
+possible, it is recommended to create the "Settle" span as a child of the
+"Deliver" span.
+
+"Settle" spans should link to creation context of the messages that are
+settled, when possible.
+
+No settlement span should be created for "fire-and-forget" scenarios, where
+messages aren't settled at all.
```

Comment:
I've had a number of comments and concerns regarding settlement within the working group meetings. I think it comes down to not clearly defining what we mean by settlement, and properly considering all types of settlement. On the topic of message settlement, I think AMQP does an excellent job of describing the 3 settlement modes that I've seen used by various messaging systems.

There's one small part of one section of the AMQP spec that I think does a good job of summing them up. It is the last part of section 2.6.12, starting with the sentence beginning with "This basic pattern...":
http://docs.oasis-open.org/amqp/core/v1.0/csprd01/amqp-core-transport-v1.0-csprd01.html#:~:text=endpoint%20to%20settle.-,This%20basic%20pattern,-can%20be%20modified

Summarizing even more quickly:
* At-most-once: Sender settles before send, no settlement on receiver.
* At-least-once: Receiver settles after process, settlement is zero duration.
* Exactly-once: Sender settles after receiver transitions a message to terminal state, receiver settles after sender settles. Receiver settlement involves a round trip exchange with the sender.

I think it important for this OTEP to provide recommendations on how to trace settlement for all of these delivery modes. Furthermore, I think it should be possible to discern from the tracing output which delivery mode is being used. Some of this may come from span structure (this OTEP) and some may come from attributes (out of scope for this OTEP), but I think we should understand what our intentions are to ensure we are making the correct recommendations here.

The two easiest are delivery modes that are fairly well described here already, at-most-once and exactly-once. I'll discuss at-least-once last.

### At-most-once Delivery
With at-most-once delivery, the message itself is fire and forget, and there's nothing to record with regards to settlement in the receiver. Since the messaging model presented in the "Stages of Producing and Consuming Messages" shows consumers receive messages from an intermediary, and intermediary instrumentation is out of scope, there is nothing to trace with regards to at-most-once delivery to consumers. I believe this is what this sentence is trying to convey:
> No settlement span should be created for "fire-and-forget" scenarios, where messages aren't settled at all.

My only comment with regards to this sentence is that "messages aren't settled at all" should perhaps say "the receiver" is not involved in settling the message delivery. Ideally "fire-and-forget" would be replaced with "at-most-once delivery":
> No settlement span should be created for at-most-once delivery scenarios, where the receiver is not involved in settling the message.

### Exactly-once Delivery
With exactly once delivery, settlement on the consumer begins when the consumer acknowledges (i.e. transitions message to a terminal state), and ends when the consumer receives confirmation of that acknowledgment from the intermediary. I think this is the delivery mode that is primarily in mind with most discussion of settlement so far in this OTEP.

### At-least-once Delivery
With at-least-once delivery, the receiver settles, and although the settlement itself is a fire-and-forget operation, recording the settlement is important, as it indicates the outcome of processing the message.

There are two scenarios we might consider with at least once delivery settlement:
1. Settlement occurs within the context of an active "Deliver" span (either due to a client ack within the context of the deliver callback, or an "auto ack" based on the callback result). In this case, how should we record settlement? This is a very common scenario, so I think it might be worth considering a special case to optimize. Requiring a zero duration "Settle" span would be unfortunate.
2. In "pull" scenarios, or in "push" scenarios where the message is not processed before the delivery callback returns, I don't see an option other than to create a zero duration "Settle" span to record the result.

Reply:
> * At-most-once: Sender settles before send, no settlement on receiver.
> * At-least-once: Receiver settles after process, settlement is zero duration.
> * Exactly-once: Sender settles after receiver transitions a message to terminal state, receiver settles after sender settles. Receiver settlement involves a round trip exchange with the sender.

The OTEP proposes to create a "Settle" span for at-least-once and exactly-once scenarios, as those are the scenarios for which any settlement operation happens on the consumer side. I think this conceptually makes sense.

I'm not sure if it is correct to say that "Settle" spans in the at-least-once scenario are zero-duration. It doesn't involve a round-trip exchange, but it involves the consumer sending something over the wire. As with "Create" spans, the duration might in most cases not be very interesting, however, I think it still represents an operation that can be modelled by a span.

That said, I like the terminology that you propose, and I think we should adopt it when talking and writing about settlement. I will rewrite the section accordingly.

However, I don't see substantial changes to the actual span structure we propose, which is summarized in the "Proposed changes and additions to the messaging semantic conventions" section:

> "Settle" spans SHOULD be created for every manually or automatically triggered settlement operation. A single "Settle" span can account for a single message or for multiple messages (in case messages are passed for settling as batches). For each message it accounts for, the "Settle" span MAY link to the creation context of the message.

Reply:
I still don't understand why we'd need to trace settlement (when it happens on broker) on the client.
Perhaps what we want is to record that message was delivered to a client (that we should cover with transport-level instrumentation in the future)? 

When auto-settlement on the broker is configured, this transport-level client-got-a-message event/span would be useful without any settlement semantics and it'd be confusing to have a settlement-specific one  there.

Reply:
> I still don't understand why we'd need to trace settlement, when it happens on the broker, on the client side.

We don't need to do that. As @dpauls said, "there is nothing to trace with regards to at-most-once delivery to consumers".

I tried to change the text in the settlement section to make this clearer.

Reply:
Based on discussions in the workgroup meeting: references to the AMQP terminology (at-most-once, at-least-once, exactly-once) were removed to avoid confusion.